presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -1653,29 +1616,52 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
         }
 
         Optional<ConnectorTablePartitioning> tablePartitioning = Optional.empty();
-        if (isBucketExecutionEnabled(session) && hiveLayoutHandle.getBucketHandle().isPresent()) {
-            tablePartitioning = hiveLayoutHandle.getBucketHandle().map(hiveBucketHandle -> new ConnectorTablePartitioning(
+        if (isBucketExecutionEnabled(session) && hiveTable.getBucketHandle().isPresent()) {
+            tablePartitioning = hiveTable.getBucketHandle().map(bucketing -> new ConnectorTablePartitioning(
                     new HivePartitioningHandle(
-                            hiveBucketHandle.getReadBucketCount(),
-                            hiveBucketHandle.getColumns().stream()
+                            bucketing.getReadBucketCount(),
+                            bucketing.getColumns().stream()
                                     .map(HiveColumnHandle::getHiveType)
-                                    .collect(Collectors.toList()),
+                                    .collect(toImmutableList()),
                             OptionalInt.empty()),
-                    hiveBucketHandle.getColumns().stream()
+                    bucketing.getColumns().stream()
                             .map(ColumnHandle.class::cast)
                             .collect(toList())));
         }
 
-        return new ConnectorTableLayout(
-                hiveLayoutHandle,
-                Optional.empty(),
+        return new ConnectorTableProperties(
                 predicate,
                 tablePartitioning,
                 Optional.empty(),
                 discretePredicates,
                 ImmutableList.of());
     }
 
+    @Override
+    public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle tableHandle, Constraint constraint)
+    {
+        HiveTableHandle handle = (HiveTableHandle) tableHandle;
+        checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
+
+        if (handle.getPartitions().isPresent()) {
+            return Optional.empty(); // TODO: optimize multiple calls to applyFilter
+        }
+
+        HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, handle, constraint);
+        ConstraintApplicationResult<ConnectorTableHandle> result = new ConstraintApplicationResult<>(
```

Comment:
Assign to `newHandle` directly and move this to the end of the method. So:

```java
HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult);

if (handle.getPartitions().equals(newHandle.getPartitions()) &&
        handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate()) &&
        handle.getBucketFilter().equals(newHandle.getBucketFilter())) {
    return Optional.empty();
}

return Optional.of(new ConstraintApplicationResult<>(newHandle, partitionResult.getUnenforcedConstraint()));
```

Reply:
Good idea, this is better